Second version of linalg_ext.scan

harsh-nod opened the pull request:
This patch modifies the linalg_ext.scan op to
reuse the initial passed identity tensor to contain
the most recent result of the scan operation. This is
to enable a more composable transform during tiling.

Furthermore, this op no longer accepts scalars and
hence scalars must be expressed as rank-0 tensors
or memrefs.

TEST: Added tests in convert_to_loops.mlir, tiling.mlir
and e2e tests in scan.mlir.

MaheshRavishankar commented on iree/test/e2e/linalg_ext_ops/scan.mlir:
```diff
@@ -2,65 +2,80 @@ func @scan_1d_dim0_inclusive_sum() {
   %init = linalg.init_tensor [6] : tensor<6xf32>
-  %c0 = arith.constant 0.0 : f32
-  %0 = iree_linalg_ext.scan
+  %t0 = util.unfoldable_constant dense<0.0> : tensor<f32>
```
Maybe have a non-zero value here so that you can truly check that the initial value has been added.

harsh-nod replied:
Sure will change.

MaheshRavishankar reviewed (approved):
Very nice! THanks!

MaheshRavishankar commented:
> This patch modifies the linalg_ext.scan op to reuse the initial passed identity tensor to contain the most recent result of the scan operation. This is to enable a more composable transform during tiling.
> 
> Furthermore, this op no longer accepts scalars and hence scalars must be expressed as rank-0 tensors or memrefs.
> 
> TEST: Added tests in convert_to_loops.mlir, tiling.mlir and e2e tests in scan.mlir.

Ill merge after you let me know if you want to address the minor comment or not.

harsh-nod commented:
> > This patch modifies the linalg_ext.scan op to reuse the initial passed identity tensor to contain the most recent result of the scan operation. This is to enable a more composable transform during tiling.
> > Furthermore, this op no longer accepts scalars and hence scalars must be expressed as rank-0 tensors or memrefs.
> > TEST: Added tests in convert_to_loops.mlir, tiling.mlir and e2e tests in scan.mlir.
> 
> Ill merge after you let me know if you want to address the minor comment or not.

I'll fix the change and let you know when its ready.

harsh-nod commented:
> Very nice! THanks!

Thanks! :)

harsh-nod commented:
> > > This patch modifies the linalg_ext.scan op to reuse the initial passed identity tensor to contain the most recent result of the scan operation. This is to enable a more composable transform during tiling.
> > > Furthermore, this op no longer accepts scalars and hence scalars must be expressed as rank-0 tensors or memrefs.
> > > TEST: Added tests in convert_to_loops.mlir, tiling.mlir and e2e tests in scan.mlir.
> > 
> > 
> > Ill merge after you let me know if you want to address the minor comment or not.
> 
> I'll fix the change and let you know when its ready.

I updated the test and everything still passes. Please merge whenever you can. Thanks!
